### DISCUSSION TITLE
More string manipulation components

### DISCUSSION THREAD

**Discussion by FalcoGer** [FakeFishGames/Barotrauma#9938] [Ideas and suggestions]
Right now string signals are very limited. You can only concatenate them and search them with regex. You can remove substrings from strings with regex by having output set and using capture groups, but it's somewhat more complicated than it needs to be.

I propose the following components:
Inputs marked with * may be preset using the interface, similar to a memory unit's initial value.

Find:
Inputs: string, substring*
Output: index
Function: Finds the substring in the string and outputs the 0 based index of the first character, -1 if not found

String length:
Inputs: string
Output: length of string
Function: Outputs the length of a string

Substring:
Inputs: string, index*, length*
Output: string
Function: outputs the substring from index to index + length. if index is negative, then count from the end of the string backwards, for example "hello world", -3, 2 would be "rl". if length is negative, reverse the output. if length overruns the end (or start) of the string, then just output up to the end/start

Edit:
Inputs: string*, regex*
Output: string
Function: Edits the string with the rules of the linux tool `sed`. For instance a substitute (replace) command looks something like this: `s/^my name is (.*)$/hello \1/i` would replace "My name is Tom" with "hello Tom", ignoring case.
Alternatively just a regex replace with capture groups and would suffice.

**Comment by mygamingaccount** [comment]
There is no use for these operations on a submarine

**Reply by FalcoGer** [reply]
Of course there is. For example I'm using circuits to get a list of all current alarms in a terminal. Something more involved than concat and regex would go a long way.

**Comment by gerakis** [comment]
If you use regex capture groups you can kind of do a substring if you know what you are looking for. The rest I agree might be helpful for heavily wired subs.

**Reply by FalcoGer** [reply]
Yes, that requires knowing some sort of delimiter and/or end. And of course it's awkward and requires jumping hoops. You could say you don't need the subtract component, you can just multiply with "-1" and add. And yet we have a subtract component.
